### ISSUE TITLE
On bottle fill

### ISSUE THREAD

**Issue by Ydda** [bensku/Skript#1070]
It would be great to have an event for filling a bottle, similar to 'on bucket fill'.

Filling of a bottle cannot really be handled by using 'on right click' as right clicking water does not register and it can't be handled by using 'on break of water' since filling a bottle does not actually change the water into air.

I hope that I'm not missing anything obvious.

**Comment by bensku** [comment]
Sounds like a bug with right click event.

**Comment by TheBentoBox** [comment]
> I hope that I'm not missing anything obvious.

Well, I guess the obvious thing is that there isn't a Spigot bottle fill event as far as I'm aware, meaning there is no event we can listen to for this. I'm not gonna close this immediately in case someone else can correct me on that but I'm pretty sure we'd have to implement it in the same janky way you'd have to, which means you should probably just do it yourself.

And on that note, you can just use target block instead of clicked block. e.g.
```
on right click:
    player's tool is glass bottle
    player's target block is water block
    distance between player and player's target block <= 5
    <event>
```
Of course this would need some tweaking, mainly with the distance check since I'm sure that's not exact, but you get the idea. Could also add a `wait a tick` then check if their tool is a filled water bottle to help make it more foolproof.

**Comment by TheBentoBox** [comment]
Also I don't believe this is a bug as water has never fired a click event since the player can't actually interact with or target it; I don't think the client event communicates to the server if it clicks on water with an open hand (unless it clicks through the water to a block behind it, of course). Needs more testing to verify but I believe that was always the case and that it's meant to be that way.

**Comment by Ydda** [comment]
I don't believe it is a bug either. I was thinking more in terms of adding a on bottle fill feature, since I am having difficulty cancelling the filling of a bottle.

**Comment by bensku** [comment]
Well, you do click on water... Not a bug, but maybe an enhancement?

**Comment by supertassu** [comment]
> Could also add a wait a tick then check if their tool is a filled water bottle to help make it more foolproof.

This would break if you had more than 1 bottle in the stack.

**Comment by kerudion** [comment]
If this helps anyone, this is what I used in one of my plugin to detect (and override) bottle filling:
```java
// PlayerInteractEvent can sometimes be already cancelled by the server itself,
// so ignoreCancelled should be false (the default)
@EventHandler
private void onInteract(PlayerInteractEvent event) {
	ItemStack item = event.getItem();
	if (item == null || item.getType() != Material.GLASS_BOTTLE || !event.getAction().isRightClick())
		return;

	Player player = event.getPlayer();

	RayTraceResult result = player.getWorld().rayTraceBlocks(player.getEyeLocation(),
		player.getEyeLocation().getDirection(), player.getAttribute(Attribute.BLOCK_INTERACTION_RANGE).getValue(),
		FluidCollisionMode.SOURCE_ONLY);
	if (result == null || result.getHitBlock() == null || result.getHitBlock().getType() != Material.WATER)
		return;

	// if you want to override the filling:
	event.setUseItemInHand(Result.DENY);

	if (player.getGameMode() != GameMode.CREATIVE)
		item.subtract();

	ItemStack filled = new ItemStack(Material.POTION);

	PotionMeta meta = (PotionMeta) filled.getItemMeta();
	meta.setBasePotionType(PotionType.WATER);

	// maybe do something with the meta

	filled.setItemMeta(meta);

	player.getInventory().addItem(filled).values().stream()
		.forEach(i -> player.getWorld().dropItemNaturally(player.getLocation(), i));
}
```
